File: src/main/java/com/braydenoneal/blang/parser/expression/operator/BooleanOperator.java
package com.braydenoneal.blang.parser.expression.operator;

import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.expression.Expression;
import com.braydenoneal.blang.parser.expression.value.BooleanValue;
import com.braydenoneal.blang.parser.expression.value.Value;

public record BooleanOperator(
        String operator,
        Expression operand_a,
        Expression operand_b
) implements Operator, Expression {
    @Override
    public Value<?> evaluate(Program program) {
        Value<?> a = operand_a.evaluate(program);
        Value<?> b = operand_b.evaluate(program);

        if (a instanceof BooleanValue a1 && b instanceof BooleanValue b1) {
            if (operator.equals("and")) {
                return new BooleanValue(a1.value() && b1.value());
            } else {
                return new BooleanValue(a1.value() || b1.value());
            }
        }

        System.out.println("boolean operator");
        System.out.println(operator);
        System.out.println(operand_a);
        System.out.println(operand_b);
        return null;
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/operator/BooleanOperator.kt
package com.braydenoneal.blang.parser.expression.operator

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.RunException
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.braydenoneal.blang.parser.expression.value.BooleanValue
import com.braydenoneal.blang.parser.expression.value.Value
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder

data class BooleanOperator(
    val operator: String,
    val operandA: Expression,
    val operandB: Expression
) : Operator, Expression {
    override fun evaluate(program: Program): Value<*> {
        val a = operandA.evaluate(program)
        val b = operandB.evaluate(program)

        if (a is BooleanValue && b is BooleanValue) {
            if (operator == "and") {
                return BooleanValue(a.value && b.value)
            }

            return BooleanValue(a.value || b.value)
        }

        throw RunException("Operands are not booleans")
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.BOOLEAN_OPERATOR

    companion object {
        val CODEC: MapCodec<BooleanOperator> = RecordCodecBuilder.mapCodec {
            it.group(
                Codec.STRING.fieldOf("operator").forGetter(BooleanOperator::operator),
                Expression.CODEC.fieldOf("operand_a").forGetter(BooleanOperator::operandA),
                Expression.CODEC.fieldOf("operand_b").forGetter(BooleanOperator::operandB)
            ).apply(it, ::BooleanOperator)
        }
    }
}
